src/main/kotlin/com/cognifide/gradle/aem/common/instance/InstanceIMSClient.kt

```diff
@@ -0,0 +1,161 @@
+
+    private lateinit var clientId: String
+
+    private lateinit var clientSecret: String
+
+    private lateinit var imsHost: String
+
+    private lateinit var imsExchangeEndpoint: String
+
+    private val expirationTime = System.currentTimeMillis() / 1000 + 86400L
+
+    private lateinit var metaScopes: List<String>
+
+    private fun readProperties() {
+        if (serviceTokenFile == null) {
+            throw InstanceException("No secret file available for generating access token")
+        }
+        val jsonString = Files.readString(serviceTokenFile!!.toPath())
+        val obj = JSONObject(jsonString)
+        val integration = obj.getJSONObject("integration")
+
+        privateKey = integration.getString("privateKey")
+        orgId = integration.getString("org")
+        technicalAccountId = integration.getString("id")
+        clientId = integration.getJSONObject("technicalAccount").getString("clientId")
+        clientSecret = integration.getJSONObject("technicalAccount").getString("clientSecret")
+        imsHost = integration.getString("imsEndpoint")
+        imsExchangeEndpoint = "https://$imsHost/ims/exchange/jwt"
+        metaScopes = integration.getString("metascopes").split(",")
+    }
+
+    private fun getJWTToken(): String {
+        var privateKeyContent: ByteArray
+        PEMParser(StringReader(privateKey)).use { pemParser ->
+            val keyPair: KeyPair = JcaPEMKeyConverter().getKeyPair(pemParser.readObject() as PEMKeyPair)
+            privateKeyContent = keyPair.private.encoded
+        }
+        val keyFactory = KeyFactory.getInstance("RSA")
+        val keySpec: KeySpec = PKCS8EncodedKeySpec(privateKeyContent)
+        val rsaPrivateKey = keyFactory.generatePrivate(keySpec) as RSAPrivateKey
+
+        val jwtClaims: MutableMap<String, Any> = HashMap()
+        jwtClaims["iss"] = orgId
+        jwtClaims["sub"] = technicalAccountId
+        jwtClaims["exp"] = expirationTime
+        jwtClaims["aud"] = "https://$imsHost/c/$clientId"
+        for (scope in metaScopes) {
+            jwtClaims["https://$imsHost/s/$scope"] = true
+        }
+
+        return Jwts.builder()
+            .setClaims(jwtClaims)
+            .signWith(SignatureAlgorithm.RS256, rsaPrivateKey)
+            .compact()
+    }
+
+    private fun getAccessToken(jwtToken: String): String? {
+        val connection = URL(imsExchangeEndpoint).openConnection() as HttpsURLConnection
+        connection.requestMethod = HttpMethods.POST
+        val urlParameters = "client_id=$clientId&client_secret=$clientSecret&jwt_token=$jwtToken"
+
+        // Send post request
+        connection.doOutput = true
+        val writer = DataOutputStream(connection.outputStream)
+        writer.writeBytes(urlParameters)
+        writer.flush()
+        writer.close()
+
+        var responseError = false
+
+        val inputStream: InputStream
+        if (connection.responseCode < HttpsURLConnection.HTTP_BAD_REQUEST) {
```

Comment:
looks risky / not so strict

Reply:
I got this one from adobe repo, they use it this way, I also was skeptical about it but left it as is, anyway changed now